feat: реализован MapReduce фреймворк

FistMr opened the pull request:
Основные компоненты:
- Coordinator - управляет задачами и worker'ами
- Worker - выполняет map/reduce задачи
- WordCount - пример реализации map/reduce функций

Ключевые возможности:
- Параллельная обработка map задач
- Группировка и сортировка данных для reduce
- Запись промежуточных и финальных результатов
- Логирование на всех этапах работы

Для запуска: java MapReduce <файлы> <воркеры> <reduce задачи>
Пример: java MapReduce file1.txt file2.txt 2 1
Добавлены тестовые файлы в корне:
  - file1.txt
  - file2.txt

FistMr committed:
feat: реализован MapReduce фреймворк

Основные компоненты:
- Coordinator - управляет задачами и worker'ами
- Worker - выполняет map/reduce задачи
- WordCount - пример реализации map/reduce функций

Ключевые возможности:
- Параллельная обработка map задач
- Группировка и сортировка данных для reduce
- Запись промежуточных и финальных результатов
- Логирование на всех этапах работы

Для запуска: java MapReduce <файлы> <воркеры> <reduce задачи>
Пример: java MapReduce file1.txt file2.txt 2 1
Добавлены тестовые файлы в корне:
  - file1.txt
  - file2.txt

vladislavTelesh commented on src/main/java/com/puchkov/Coordinator.java:
```diff
@@ -0,0 +1,99 @@
+    }
+
+    public synchronized Task getTask() {
+
```
если не ошибаюсь, то задача помечается как выполненная только после завершения, но не помечается как "в работе" при назначении.

FistMr replied:
Сделал Enum Status вместо boolean

vladislavTelesh commented on src/main/java/com/puchkov/Worker.java:
```diff
@@ -0,0 +1,111 @@
+
+    @Override
+    public void run() {
+        while (true) {
```
Если координатор перестаёт выдавать задачи, но ещё не сообщает о завершении, воркер застрянет в бесконечном цикле ?

FistMr replied:
Coordinator возвращает null если все map задачи уже разданы - в этом случае worker засыпает на некоторое время, чтобы дождаться завершения всех map задач, 
Если Все задачи выполнены, то Coordinator вернет Task.Type.NONE, что заставит Worker-a завершить работу.

FistMr replied:
добавил проверку, что координтор живой через таймер выполнения последнего задания, с аварийным завершением воркера.

vladislavTelesh commented on src/main/java/com/puchkov/Worker.java:
```diff
@@ -0,0 +1,111 @@
+
+                    coordinator.completeTask(task);
+                } catch (IOException e) {
+                    logger.error("Ошибка обработки MAP-задачи {}: {}", task.taskId, e.getMessage());
```
я бы что-то тип такого добавил coordinator.reportTaskFailure(task); ещё

FistMr replied:
добавил метод coordinator.reportTaskFailure(task); который возвращает статус таски на AWAITING в случае если worker упал или не сделал таску

vladislavTelesh commented on src/main/java/com/puchkov/Coordinator.java:
```diff
@@ -0,0 +1,99 @@
+import java.util.concurrent.atomic.AtomicInteger;
+
+class Coordinator {
+    private static final Logger logger = LoggerFactory.getLogger(Coordinator.class);
```
можно добавить ещё таймауты для выполнения задач, чтобы задачи не зависали

FistMr replied:
сделал вложенный статичный класс TaskInfo, в котором содержится информация о времени выполнения и статусе таски, когда все map или reduce задачи разданы, свободные воркеры будут проверять таски на зависание в методе checkForStuckTasks

vladislavTelesh commented on src/main/java/com/puchkov/Worker.java:
```diff
@@ -0,0 +1,111 @@
+import org.slf4j.LoggerFactory;
+
+import java.io.*;
+import java.nio.file.Files;
```
а очистка временных файлов есть ?

FistMr replied:
добавил метод cleanIntermediateFiles в Coordinator

vladislavTelesh commented on src/main/java/com/puchkov/Worker.java:
```diff
@@ -0,0 +1,111 @@
+
+                    List<KeyValue> kvs = mapFunc.map(task.inputFile, content);
+
+                    Map<Integer, List<KeyValue>> buckets = new HashMap<>();
```
а не ConcurrentHashMap лучше ?

FistMr replied:
В данном случае buckets используется в рамках только одного потока, поэтому ConcurrentHashMap скорее всего тут только замедлит запись/чтение

vladislavTelesh commented on src/main/java/com/puchkov/Coordinator.java:
```diff
@@ -0,0 +1,99 @@
+        }
+    }
+
+    public synchronized Task getTask() {
```
как будто не оч эффективно 
мб синхронизацию использовать для map задач и для reduce задач только если все map-задачи завершены ?

FistMr replied:
сделал getTask не синхронизированным, добавив условия проверки количеств выполненных map задач.Вынес логику выдачи map и reduce задач в отдельные synchronized методы

FistMr committed:
refactor: переработал управление задачами и ресурсами

Основные изменения:
- Заменил boolean статусы на Enum в классе TaskInfo
- Реализовал автоматическую очистку временных файлов:
- Добавил механизм таймаутов:
  - 30-секундный таймаут для зависших задач
  - Аварийный выход воркеров при долгом ожидании
- Ввел обработку сбоев через reportTaskFailure()
- Оптимизировал синхронизацию:
  - getTask() теперь несинхронизированный
  - Логика выдачи задач в отдельных sync-методах
- Улучшил проверку завершения MAP-фазы
- Убрал избыточные логи

FistMr committed:
add README.md
